macrosynergy/management/utils/df_utils.py

```diff
@@ -300,6 +300,65 @@ def downsample_df_on_real_date(
     )
 
 
+def downsample_wide_df_on_real_date(
+    df: pd.DataFrame,
+    freq: str = "M",
+    agg: str = "mean",
+    ffill: bool = False,
```

Comment:
If the function ifs for down-sampling, we should never need to forward or backward fill.

Reply:
the fills exist to pull data from whichever direction we will next use it if missing on an index.
e.g. If I downsample a daily dataframe to a monthy frequency, I end up with data only on the end of the month (30-Jan for example). 
Now when I want to use this data with weights, or in any method where I must consider the value for the next/previous data available data point, it leads to very large number of calls to pandas slicing/iloc/first_valid/last_valid_index

Reply:
That seems to me a question of two part in the data-frame to have the frequency converted:

1. All non-last period (of down-sampling): so for monthly output targets, we can take everything before the last month and directly change frequency.
2. The last month in the sample, where we need to take the last observation, but we have two options: stock or flow variable. Stock variable is equivalent to the assumption of forward fill (last value), where as a flow variable (such as returns), needs to be the sum up to that date.

Reply:
How do we differentiate for (2) when contract signals are provided?